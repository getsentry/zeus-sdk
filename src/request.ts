import fetch, { RequestInit, Response } from 'node-fetch';

/**
 * Request options.
 *
 * As opposed to the default fetch options, this interface only permits headers
 * to be passed as object. This is needed as headers are preprocessed by the
 * request utility.
 */
export type RequestOptions = {
  [P in keyof RequestInit]: P extends 'headers'
    ? Record<string, string>
    : RequestInit[P]
};

export { Response };

/**
 * Parses an error message from the given response.
 *
 * First, this function tries to parse an error message from the response body.
 * If this does not work, it falls back to the HTTP status text.
 *
 * @param response A fetch Response object.
 * @returns A promise that resolves the error.
 */
async function parseError(response: Response): Promise<Error> {
  try {
    const { message } = (await response.json()) as { message?: string };
    return new Error(message || `${response.status} ${response.statusText}`);
  } catch (e) {
    return new Error(`${response.status} ${response.statusText}`);
  }
}

/**
 * Performs an AJAX request to the given url with the specified options using
 * fetch.
 *
 * After the request has finished, the result is parsed and checked for errors.
 * In case of an error, the response message is thrown as an error. On success,
 * the response object is passed into the promise.
 *
 * @param url The destination of the AJAX call.
 * @param options Options to the fetch call.
 * @returns A Promise to the parsed response body.
 */
export async function request(
  url: string,
  options: RequestOptions = {}
): Promise<Response> {
  const response = await fetch(url, options);
  if (!response.ok) {
    throw await parseError(response);
  }
  return response;
}
