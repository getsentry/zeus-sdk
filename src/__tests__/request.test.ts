import fetch from 'node-fetch';
import { request } from '../request';

const fetchMock = fetch as jest.Mock<typeof fetch>;

async function mockPromise<T>(value?: T, reason?: string): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    setImmediate(() => {
      if (reason) {
        reject(reason);
      } else {
        resolve(value);
      }
    });
  });
}

function mockFetch<T>(
  status: number,
  json: () => Promise<T>,
  statusText?: string
): void {
  const ok = status >= 200 && status <= 300;
  fetchMock.mockReturnValue(mockPromise({ status, ok, json, statusText }));
}

describe('request', () => {
  beforeEach(() => {
    fetchMock.mockReset();
  });

  test('passes all parameters to fetch', async () => {
    mockFetch(200, mockPromise);
    await request('http://example.org', {
      headers: { Authorization: 'bearer yo!' },
      method: 'POST',
    });

    expect(fetchMock).toMatchSnapshot();
  });

  test('allows Accept header overrides', async () => {
    mockFetch(200, mockPromise);
    await request('http://example.org', {
      headers: { Accept: 'text/html' },
    });

    expect(fetchMock).toMatchSnapshot();
  });

  test('throws an error containing the status text', async () => {
    mockFetch(400, async () => mockPromise(undefined, 'empty'), 'BAD REQUEST');

    try {
      await request('http://example.org');
    } catch (e) {
      expect((e as Error).toString()).toMatch('400 BAD REQUEST');
    }
  });

  test('throws an error containing the resolved error message', async () => {
    const message = 'Error message';
    mockFetch(400, async () => mockPromise({ message }));

    try {
      await request('http://example.org');
    } catch (e) {
      expect((e as Error).toString()).toMatch(message);
    }
  });

  test('falls back to the status text when parsing errors', async () => {
    mockFetch(400, async () => mockPromise({}), 'BAD REQUEST');

    try {
      await request('http://example.org');
    } catch (e) {
      expect((e as Error).toString()).toMatch('400 BAD REQUEST');
    }
  });
});
