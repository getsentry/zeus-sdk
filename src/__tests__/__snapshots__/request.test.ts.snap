// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request allows Accept header overrides 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://example.org",
      Object {
        "headers": Object {
          "Accept": "text/html",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`request passes all parameters to fetch 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://example.org",
      Object {
        "headers": Object {
          "Authorization": "bearer yo!",
        },
        "method": "POST",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
